lib/player.rb

```diff
@@ -12,44 +12,116 @@ def add_ship(ship)
     @ships << ship
   end
 
+
+
   def ship_setup
     if @type == :auto
-      # helper-method: auto_generation CALLS
-      # helper-method: player_place_ships
-      # auto generation
-      p "type auto"
-    elsif @type == :user
-      # Put in a guard for user error
-
-      # helper-method: player_place_ships
-      @ships.each do |ship|
-        if ship.length == 3
-          puts "Enter the squares for the Cruiser (3 spaces):"
-          puts "> "
-        elsif ship.length == 2
-          puts "Enter the squares for the Submarine (2 spaces):"
+      until all_ships_placed?
+        auto_generate_coordinates
+      end
+    else
+      until all_ships_placed?
+        @ships.each do |ship|
+          puts "Enter the squares for the #{ship.name} (#{ship.length} spaces)"
           puts "> "
-        end
-
-        user_ship_coords = gets.chomp.split(/ /)
-        # Consider loop for user error
-        if @board.valid_placement?(ship, user_ship_coords) == true
-          @board.place(ship, user_ship_coords)
-        elsif @board.valid_placement?(ship, user_ship_coords) == false
-          loop do
-            puts "Those are invalid coordinates. Please try again:"
-            puts "> "
-            user_ship_coords = gets.chomp.split(/ /)
-
-            if @board.valid_placement?(ship, user_ship_coords) == true
-              @board.place(ship, user_ship_coords)
-              require "pry"; binding.pry
-              exit # Could be a helper-method
+          user_ship_coords = gets.chomp.upcase.split(/ /)
+          result = @board.valid_placement?(ship, user_ship_coords)
+          if result == false
+            until this_ship_placed?(ship)
+              puts "Those are invalid coordinates. Please try again: "
+              user_ship_coords2 = gets.chomp.upcase.split(/ /)
+            @board.valid_placement?(ship, user_ship_coords2)
+            @board.validated_placement
             end
+          else
+            @board.validated_placement
           end
         end
       end
     end
   end
 
+  def auto_generate_coordinates
+    @ships.each do |ship|
+      until this_ship_placed?(ship)
+        auto_coords = []
+        possible = @board.cells.keys.shuffle
+        ship.length.times do
+          auto_coords << possible[0]
+          possible.rotate!
+        end
+        @board.valid_placement?(ship, auto_coords)
+        @board.validated_placement
+      end
+    end
+  end
+
+  
+  def this_ship_placed?(ship)
+    @board.cells.values.any? do |cell|
+      cell.ship == ship
+    end
+  end
+
+  def ship_count
+    @ships.sum do |ship|
+      ship.length
+    end
+  end
+
+  def ships_on_board
+    ships_on_board = @board.cells.reduce(0) do |ships_placed_count, cell|
+      if cell[1].ship != nil
+        ships_placed_count += 1
+      end
+      ships_placed_count
+    end
+  end
+
+  def all_ships_placed?
```

Comment:
What a smart idea!
Human Speak for my brain process:
1. ship_count will call the array of ships the player has. For each ship in their array it will count the length of the ship. This will RETURN the total number of cells that a player should be filling (total length of ship and total number of ships).
2. ships_on_board will create a variable for the .reduce enumerable. Understanding the .reduce: We have called the players board, then the cells array (on player board). We are iterating over the player's board cells array using .reduce with the initial value of the memo (accumulator you've labeled ships_placed_count) equal to the integer value of 0. We are iterating over cell (btw - thanks for labeling this awesomely for readability). If the player's board cell array at index 1 [which holds the cell object instantiation] has a ship (aka not equal to nil), this will cause the ships_place_count accumulator to increase by one. If, there is no ship, it will skip the if condition and just return the ships_placed_count at the current counter. This will RETURN the total number of cells that are filled on the board.
3. Finally, all_ships_placed? will query if ship_count and ships_on_board are equivalent!

SUCH AMAZING LOGIC!!!!